pkg/controller/admissionchecks/provisioning/controller.go

```diff
@@ -833,3 +833,116 @@ func limitObjectName(fullName string) string {
+	PodSetAssignment *kueue.PodSetAssignment
+	Count            int32
+}
+
+func mergePodSets(
+	expectedPodSets []kueue.PodSetReference,
+	podSetMap map[kueue.PodSetReference]*kueue.PodSet,
+	psaMap map[kueue.PodSetReference]*kueue.PodSetAssignment,
+	mergeMode *kueue.MergePodSetsMode,
+) ([]MergedPodSet, error) {
+	merged := []MergedPodSet{}
+
+	if mergeMode == nil {
+		for _, psName := range expectedPodSets {
+			ps, psFound := podSetMap[psName]
+			psa, psaFound := psaMap[psName]
+			if !psFound || !psaFound {
+				return nil, errInconsistentPodSetAssignments
+			}
+			merged = append(merged, MergedPodSet{
+				Name:             psName,
+				PodSet:           ps,
+				PodSetAssignment: psa,
+				Count:            ptr.Deref(psa.Count, ps.Count),
+			})
+		}
+		return merged, nil
+	}
+
+	for _, psName := range expectedPodSets {
+		ps, psFound := podSetMap[psName]
+		psa, psaFound := psaMap[psName]
+		if !psFound || !psaFound {
+			return nil, errInconsistentPodSetAssignments
+		}
+
+		mergable := false
+		for i, mps := range merged {
+			if canMergePodSets(mps.PodSet, ps, mergeMode) {
+				merged[i].Count += ptr.Deref(psa.Count, ps.Count)
+				mergable = true
+				break
+			}
+		}
+
+		if !mergable {
+			merged = append(merged, MergedPodSet{
+				Name:             psName,
+				PodSet:           ps,
+				PodSetAssignment: psa,
+				Count:            ptr.Deref(psa.Count, ps.Count),
+			})
+		}
+	}
+
+	return merged, nil
+}
+
+func canMergePodSets(ps1, ps2 *kueue.PodSet, mergeMode *kueue.MergePodSetsMode) bool {
+	switch *mergeMode {
+	case kueue.Identical:
+		// Strict comparison: all fields must match
+		return arePodSetsIdentical(ps1, ps2)
+	case kueue.Equivalent:
+		// Relaxed comparison: only resource requests must match
+		return arePodSetsEquivalent(ps1, ps2)
+	default:
+		// Unsupported merge mode
+		return false
+	}
+}
+
+func arePodSetsIdentical(ps1, ps2 *kueue.PodSet) bool {
+	return equality.Semantic.DeepEqual(ps1.Template, ps2.Template)
+}
+
+func arePodSetsEquivalent(ps1, ps2 *kueue.PodSet) bool {
+	ps1ContainersRes := aggregateTotalResources(ps1.Template.Spec.Containers)
+	ps2ContainersRes := aggregateTotalResources(ps2.Template.Spec.Containers)
+	areContainersEqual := equality.Semantic.DeepEqual(ps1ContainersRes, ps2ContainersRes)
+
+	ps1InitContainersRes := aggregateTotalResources(ps1.Template.Spec.InitContainers)
+	ps2InitContainersRes := aggregateTotalResources(ps2.Template.Spec.InitContainers)
+	areInitContainersEqual := equality.Semantic.DeepEqual(ps1InitContainersRes, ps2InitContainersRes)
+
+	return areContainersEqual &&
+		areInitContainersEqual &&
+		equality.Semantic.DeepEqual(ps1.Template.Spec.Resources, ps2.Template.Spec.Resources) &&
+		equality.Semantic.DeepEqual(ps1.Template.Spec.NodeSelector, ps2.Template.Spec.NodeSelector) &&
+		equality.Semantic.DeepEqual(ps1.Template.Spec.Tolerations, ps2.Template.Spec.Tolerations) &&
+		equality.Semantic.DeepEqual(ps1.Template.Spec.Affinity, ps2.Template.Spec.Affinity) &&
```

Comment:
Do you really want to contain the `affinity` in identical keys? The affinity is a bit of a flexible field. So, if we contain this in here, I can imagine this identical check is much stricter. In other words, this affinity inclusion blocks PodSet merges in a lot of cases.

Reply:
I think it is possible that ClusterAutoscaler is using Affinities for deciding on the structure of the newly provisioned nodes. So, some users may not want to have the same "merged" provisioning semantics. For example, if Pods from one Master have "anti-affinity" to pods in Workers. 

AFAIK we don't have a use-case currently that would discriminate the semantics. Maybe in the future we need more knobs to decide what "Equivalent/Similar" means.
 
 wdyt @mwielgus ?